src/db.c

```diff
@@ -48,6 +48,36 @@
 int expireIfNeeded(redisDb *db, robj *key, int flags);
 int keyIsExpired(redisDb *db, robj *key);
 
+/* Returns next dictionary from the iterator, or NULL if iteration is complete. */
+dict *dbNextDict(dbIterator *dbit) {
+    int64_t slot;
+    if (intsetGet(dbit->db->owned_slots, dbit->index++, &slot)){
+       return dbit->db->dict[slot];
+    }
+    return NULL;
+}
+
+/* Returns DB iterator that can be used to iterate through sub-dictionaries.
+ * Primary database contains only one dictionary when node runs without cluster mode,
+ * or 16k dictionaries (one per slot) when node runs with cluster mode enabled. */
+void dbInitIterator(dbIterator *dbit, redisDb *db) {
+    dbit->db = db;
+    dbit->index = 0;
+}
+
+/* Returns next non-empty dictionary strictly after provided slot and updates slot id in the supplied reference.
+ * This function doesn't use dbIterator in order to provide deterministic iteration across all owned slots. */
+dict *dbGetNextUnvisitedSlot(redisDb *db, int *slot) {
+    for (int i = *slot + 1; i < db->dict_count; i++) {
+        if (!dictIsEmpty(db->dict[i])) {
```

Comment:
The comment mentions "iteration across all owned slots" but it actually includes non-owned slots (e.g. importing slots not yet owned). What do we want here, owned non-empty or all non-empty?

Reply:
I think we want owned AND non-empty, however I believe that existing implementation of SCAN API in the unstable branch doesn't check for owned vs not owned either (this is why I've ignored check here). We should fix that perhaps.